src/playerlist.rs

```diff
@@ -0,0 +1,109 @@
+use std::{collections::HashMap, path::PathBuf};
+
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use steamid_ng::SteamID;
+
+use crate::settings::{ConfigFilesError, Settings};
+
+// PlayerList
+
+#[derive(Serialize, Deserialize)]
+pub struct Playerlist {
+    #[serde(skip)]
+    path: PathBuf,
+    pub records: HashMap<SteamID, PlayerRecord>,
+}
+
+impl Playerlist {
+    /// Attempt to load the [Playerlist] from the default file
+    pub fn load() -> Result<Playerlist, ConfigFilesError> {
+        Self::load_from(Self::locate_playerlist_file()?)
+    }
+
+    /// Attempt to load the [Playerlist] from the provided file
+    pub fn load_from(path: PathBuf) -> Result<Playerlist, ConfigFilesError> {
+        let contents = std::fs::read_to_string(&path)
+            .map_err(|e| ConfigFilesError::IO(path.to_string_lossy().into(), e))?;
+        let mut playerlist: Playerlist = serde_json::from_str(&contents)
+            .map_err(|e| ConfigFilesError::Json(path.to_string_lossy().into(), e))?;
+        playerlist.path = path;
+        Ok(playerlist)
+    }
+
+    /// Attempt to save the [Playerlist] to the file it was loaded from
+    pub fn save(&self) -> Result<(), ConfigFilesError> {
+        let contents = serde_json::to_string(self).context("Failed to serialize playerlist.")?;
+        std::fs::write(&self.path, contents)
+            .map_err(|e| ConfigFilesError::IO(self.path.to_string_lossy().into(), e))?;
+        Ok(())
+    }
+
+    fn locate_playerlist_file() -> Result<PathBuf, ConfigFilesError> {
+        Settings::locate_config_directory().map(|dir| dir.join("playerlist.json"))
+    }
+}
+
+impl Default for Playerlist {
+    fn default() -> Self {
+        let path = Self::locate_playerlist_file()
+            .map_err(|e| tracing::warn!("Failed to create config directory: {:?}", e))
+            .unwrap_or("playerlist.json".into());
```

Comment:
Does this just create a `playerlist.json` file in the same directory as the executable if we fail to find and create a config directory? Should we have a fallback in this case to create a whole `config` subdir in the same dir as the executable to house this player list and any other config files that may need to be generated? OR  am I misinterpretting that error?

Reply:
Yes you are interpreting that correctly. Ideally this should never happen, but I figured trying to put the file *somewhere* (like in the same folder as the binary) would be preferable to just automatically losing any data you might have gathered. 

I personally don't think it's very necessary to make a subdir for configs and stuff as the settings are not saved elsewhere on failure so it would just be the json file. If you would like to try fallback to another directory I think that would be out of the scope of this MR and maybe should go in a separate one. However I wouldn't spend too much time trying to fallback on failure as I don't like the odds of a fallback succeeding if the initial attempt didn't succeed.